[DS-3135] Replace log4j 1.2, which is EOL

mwoodiupui opened the pull request:
Fixes #6490
Continuation of the work begun in #1981.
CLI and webapp.s now logging properly.  All unit and integration tests succeed.

mwoodiupui committed:
[DS-3135] log4j 2 needs new configuration -- start with the command line tools

Commit:
Update POMs to version 2.10.0

Commit:
Update POMs to version 2.10.0

Commit:
Convert getLogger() and Log4J imports for Log4J 2

Commit:
Convert getLogger() and Log4J imports for Log4J 2

Commit:
Add Log4J LogManager to EventListenerExample

Commit:
Remove code related to log4j1

Commit:
Remove trailing whitespace in EventListenerExample

Commit:
Refactor import statements

Commit:
Remove bean declaration for log4v1 service

Commit:
Refactor additional loggers for log4jv2

Commit:
Refactor additional loggers for log4jv2

Commit:
Fix reference to log4jv2

Commit:
Fix dependency convergence and use POM property for log4j version

Commit:
Additional refactoring for log4jv2

Commit:
Additional refactoring for log4jv2

Commit:
Fix files not conforming to code style

Commit:
Fix import statement

Commit:
Fix dependency convergence

Commit:
Change to version 2.6.2 of log4j
To ensure compatibility with Spring Boot v1.4.4

mwoodiupui committed:
[DS-3135] Fix new class that uses old log4j.

mwoodiupui committed:
[DS-3135] More dependency fixes.

mwoodiupui committed:
[DS-3135] More dependency cleanup

mwoodiupui committed:
[DS-3135] Correct Log4J configuration file name.

mwoodiupui committed:
[DS-3135] Tell Log4J where to find its configuration.

mwoodiupui commented:
#2235 will be helpful here.  A Servlet 3+ webapp. should auto-start Log4j2 if the right JARs are included.

mwoodiupui committed:
[DS-3135] Still more dependency cleanup.

mwoodiupui committed:
[DS-3135] Squash one more log4j1 dependency.

mwoodiupui committed:
[DS-3135] Satisfy Checkstyle by removing unused includes; tighten up the code a little.

mwoodiupui commented:
Rebased yesterday on latest master.  First logging from webapp.s seen.  Now to make sure that they *all* are logging.

mwoodiupui committed:
[DS-3135] Don't configure Solr with ConfigureLog4jListener which no longer exists.

mwoodiupui committed:
[DS-3135] Finish converting log4j2 configuration; make it work for Solr; tweak Ant usage.

mwoodiupui commented:
All webapp.s are logging except spring-rest.

mwoodiupui committed:
[DS-3135] Clean out temporary exclusions; tidy indentation.

mwoodiupui committed:
[DS-3135] Set the log4jConfiguration context parameter properly.

mwoodiupui commented:
Rebased to cure conflict.

mwoodiupui commented:
Sorry, missed a conflict, rebased again.

mwoodiupui committed:
[DS-3135] Squash another unused import for Checkstyle.

mwoodiupui committed:
[DS-3135] Let Spring configure Log4J; correct logging config in Ant script.

mwoodiupui commented:
spring-rest is also logging now.  Ready for review.

tdonohue commented on dspace-api/src/main/java/org/dspace/eperson/SubscribeCLITool.java:
```diff
@@ -26,8 +26,8 @@
-import org.apache.log4j.Level;
-import org.apache.log4j.Logger;
+import org.apache.logging.log4j.Logger;
+// Replace with new log4j 2: import org.apache.logging.log4j.Level;
```
Wondering if this should be removed or replaced?

mwoodiupui replied:
I think it's a leftover comment.

tdonohue commented on dspace-api/src/main/java/org/dspace/eperson/SubscribeCLITool.java:
```diff
@@ -316,7 +316,7 @@ public static void main(String[] argv) {
 
         if (test) {
-            log.setLevel(Level.DEBUG);
+            //TODO Replace with log4j2: org.apache.logging.log4j.core.config.Configurator.setLevel();
```
Looks like this relates to previous comment, and we seem to not have updated the old code here?

mwoodiupui replied:
I think we should abjure such black magic and just drop this whole `if`.

tdonohue commented on dspace-rest/src/main/webapp/WEB-INF/web.xml:
```diff
@@ -76,6 +76,14 @@
+        <param-value>${dspace.dir}/config/log4j2.xml</param-value>
+        <description>
+	    The location of the Log4J configuration
+        </description>
```
I *think* Servlet API 3.1 requires the `<description>` to be *first* in `<context-param>` (IIRC I had to reorder some old context-params when updating to 3.1).  You may want to double check all the updated `web.xml` files to be certain they are "valid".

mwoodiupui replied:
Correct.

tdonohue commented on dspace-sword/src/main/java/org/purl/sword/client/ClientFactory.java:
```diff
@@ -7,7 +7,7 @@
 package org.purl.sword.client;
 
-import org.apache.log4j.PropertyConfigurator;
+// Replace with log4j 2: import org.apache.logging.log4j.PropertyConfigurator;
```
Commented out code that needs replaced

mwoodiupui replied:
Needs removed.

tdonohue commented on dspace-sword/src/main/java/org/purl/sword/client/ClientFactory.java:
```diff
@@ -24,8 +24,8 @@ public ClientFactory() {
-        PropertyConfigurator.configure(this.getClass().getClassLoader()
-                                           .getResource(ClientConstants.LOGGING_PROPERTY_FILE));
+        // Replace with log4j 2: PropertyConfigurator.configure(this.getClass()
+        // .getClassLoader().getResource(ClientConstants.LOGGING_PROPERTY_FILE));
```
Commented out code that needs replaced

mwoodiupui replied:
I think we shouldn't be tinkering with this at all.  Just let Log4J2 find the configuration on the classpath via its normal search.  If not found, you get basic console logging.

tdonohue replied:
That seems reasonable. If this isn't really needed, let's simply remove it.

tdonohue commented on dspace/bin/dspace:
```diff
@@ -35,13 +35,14 @@ if [ "$1" = "classpath" ]; then
-java $JAVA_OPTS -classpath $FULLPATH org.dspace.app.launcher.ScriptLauncher "$@"
+java $JAVA_OPTS \
+     -classpath $FULLPATH \
+     -Dlog4j.configurationFile=$DSPACEDIR/config/log4j2.xml \
```
Is there a way to not "hardcode" this into our commandline scripts, and simply read it from `log.init.config` setting in our dspace.cfg?

mwoodiupui replied:
Maybe just leave it unconfigured and let basic logging appear on the console?

mwoodiupui replied:
I fear that if we try to use the DSpace configuration to initialize logging, which is needed by the DSpace configuration code, we'll wind up again in a strange loop that tries to reconfigure logging after a bunch of it has already happened.  Something as simple as create-administrator in v6 pours out a long screed of Hibernate's muttering before settling down to work.

There are good reasons for logging to be entirely separate from the logged code.

tdonohue replied:
Removing the `-D` altogether sounds fine to me (if that works).  I just wasn't sure we wanted a path hardcoded in these scripts.

tdonohue commented on dspace/bin/dspace.bat:
```diff
@@ -39,12 +39,12 @@ REM Otherwise, default Java to using 256MB of memory
 
 REM Execute Java
-java %JAVA_OPTS% -classpath "%DSPACE_CLASSPATH%" org.dspace.app.launcher.ScriptLauncher %*
+java %JAVA_OPTS% -classpath "%DSPACE_CLASSPATH%" -Dlog4j.configurationFile=%DSPACEDIR%/config/log4j2.xml org.dspace.app.launcher.ScriptLauncher %*
```
Is there a way to not "hardcode" this into our commandline scripts, and simply read it from `log.init.config` setting in our dspace.cfg?

Also, not sure if this change will work if `%DSPACEDIR%` has spaces...we may need to surround the path with quotes (like `-classpath` is).

tdonohue reviewed (commented):
@mwoodiupui : gave this a thorough code review (no test yet).  The vast majority of the changes are obvious ones (changin imports & POM dependencies).  I *did* however notice some commented out "TODOs" (see below) and a few areas with (seemingly minor) questions.  I don't think any of these necessarily block moving this forward, but we might want to quickly do cleanup.

Also, in several of the POM-level, some odd alignment now appears... I didn't point this out directly inline (as it's not a blocker), but not sure if this is from you (or the original PR).  It looks like there's some accidental changes from spaces to tabs (or similar).

Overall though, this is looking good! We just need to give it another quick test.

mwoodiupui committed:
[DS-3135] Remove questionable loglevel fiddling; conform to webapp descriptor schema.

mwoodiupui committed:
Remove system property pointing to logging config.  Instead just let
framework do its default search, which should find the config. on the
classpath.

mwoodiupui committed:
[DS-3135] Remove commented-out attempt by logged code to tell logging framework its configuration path.

tdonohue commented on dspace-sword/src/main/java/org/purl/sword/client/ClientFactory.java:
```diff
@@ -24,8 +22,6 @@ public ClientFactory() {
         // configure the logger from the property file. The GUI client will
         // reload these properties if it is set to capture the output and
         // display it in a panel.
```
@mwoodiupui : It looks like we should remove this entire comment section too, since we removed the code that it refers to.

mwoodiupui replied:
Yes, indeed.  Actually, we can delete the entire surrounding method, since it's an empty 0-argument constructor.  Done.

tdonohue reviewed (approved):
👍 Looks great to me. I gave this a thorough code review (see previous review).  I also tested this lightly and all the webapps seem to be logging properly.

I think this looks good enough to merge. If we notice any minor issues in ongoing testing, we can fix them in a future PR.

mwoodiupui committed:
[DS-3135] Remove leftover comment (and surrounding empty niladic constructor).

terrywbrady reviewed (approved):
For each instance where a new pom.xml exclusion has been defined, consider adding a comment explaining the need for the exclusion.  This might be useful in the next round of library upgrades.

Tedious suggestion: consider consistently importing the LogManager class in in file that it is used.  This is primarily a cosmetic recommendation for consistency.

If my slf4j comment does not seem useful, feel free to ignore it.  I suppose I have a question why some of our modules need that flavor of the library.

Nice work!

mwoodiupui commented:
Rebased and force-pushed to cure a number of conflicts.

mwoodiupui commented on dspace-api/src/main/java/org/dspace/submit/step/DescribeStep.java:
```diff
@@ -16,6 +17,6 @@
      * log4j logger
      */
-    private static Logger log = Logger.getLogger(DescribeStep.class);
+    private static final Logger log = LogManager.getLogger();
```
No longer required.  Log4j2 will figure it out.

mwoodiupui commented on dspace-oai/pom.xml:
```diff
@@ -190,12 +206,20 @@
         <dependency>
-            <groupId>org.slf4j</groupId>
-            <artifactId>slf4j-log4j12</artifactId>
+            <groupId>org.apache.logging.log4j</groupId>
```
SLF4J is a transitive dependency of Jena.

mwoodiupui commented on dspace-services/pom.xml:
```diff
@@ -140,13 +140,20 @@
         </dependency>
+	<dependency>
+	    <groupId>org.apache.logging.log4j</groupId>
+	    <artifactId>log4j-slf4j-impl</artifactId>
```
dspace-services uses SLF4J exclusively.  Now that we have upgraded to Log4J v2 we can convert to that and remove some layers of logging frameworks.

I've been using SLF4J wherever I needed parameterized messages, since it's more efficient than multiple String concatenations.  Log4J2 has an even better version that accepts lambdas as message parameters.  I plan to use this quite a bit.

mwoodiupui commented:
@terrywbrady If you are referring to the use of fully-qualified class names rather than imports in some places, that is because DSpace *also* has a LogManager class and some code uses both.  We can't import both.

If this is too bothersome, I might say that the DSpace LogManager doesn't really *manage* anything and might be renamed LogHelper or the like.  Then we can import the Log4J LogManager wherever we need it.

tdonohue commented:
Gave this another review after the rebase.  Still looks good, and this sits at +2.   So, I'm going to go ahead and merge this.

If we notice any logging issues through further testing, we can resolve those in separate PRs.  Thanks again @mwoodiupui (and @PTrottier, who started the effort in #1981)!  Glad to see this make it to completion & approval.
